cm_sdk/models/__init__.py

```diff
@@ -0,0 +1,266 @@
+from enum import Enum
+import json
+from json import JSONEncoder
+import re
+import logging
+
+__author__ = 'e0d'
+
+
+class StringUtil(object):
+
+    class Case(Enum):
+        CAMEL = 1
+        SNAKE = 2
+
+    # credit
+    # http://stackoverflow.com/questions/19053707/convert-snake-case-snake-case-to-lower-camel-case-lowercamelcase-in-python
+    @classmethod
+    def to_camel_case(cls, snake_str):
+        components = snake_str.split('_')
+        return components[0].lower() + "".join(x.title() for x in components[1:])
+
+    @classmethod
+    def to_snake_case(cls, camel):
+        regex = re.compile(r'([A-Z])')
+        return regex.sub(r'_\1', camel).lower()
+
+    @classmethod
+    def convert_dict_keys(cls, dict_in, target_case=None):
+        _renamed = {}
+
+        if target_case == cls.Case.CAMEL:
+            converter = lambda name: cls.to_camel_case(name)
+            pass
+        elif target_case == cls.Case.SNAKE:
+            converter = lambda name: cls.to_snake_case(name)
+        else:
+            raise NotImplemented("Unsupported target case, {0}, provided.".format(target_case))
+
+        for k, v in dict_in.items():
+            _renamed[converter(k)] = v
+
+        return _renamed
+
+
+class CloudManagerJSONEncoder(JSONEncoder):
+    """
+    A custom JSON encoder to provide transparent conversion between
+    CloudManager JSON format and the Python classes that represent API
+    objects in the SDK.
+    """
+
+    def default(self, o):  # pylint: disable=method-hidden
+        if isinstance(o, CloudManagerEnum):
+            return o.to_json()
+        elif hasattr(o, "state"):
+            # converts class parameter names from snake case to camel case expected
+            # by the API and prunes nulls
+            return dict((StringUtil.to_camel_case(k), v) for k, v in o.state.iteritems() if v)
+        else:
+            return None
+
+
+class CloudManagerBase(object):
+
+    """
+    Base class for python representations of CloudManager API objects that provides
+    utility methods for static initialization, factors out common attributes and
+    provides common API object equivalence checks.
+    """
+    # All API objects have these attributes
+    COMMON_API_ATTRIBUTES = ['id', 'group_id', 'links', 'created']
+    # The following attributes are ignored for determine whether instances are equivalent.
+    # This is useful when determining if a non-persitant entity duplicates one that was
+    # created previously.
+    EQUIVALENCE_IGNORED_ATTRIBUTES = ['id', 'group_id', 'links', 'created', 'updated']
+
+    def __init__(self, my_api_attributes):
+
+        all_api_attributes = list(self.COMMON_API_ATTRIBUTES)
+        all_api_attributes.extend(my_api_attributes)
+        vars(self).update(all_api_attributes=all_api_attributes, state={})
```

Comment:
The proper way to do this is to call `super(self.__class__, self).__init__()` in the child class. It's almost never a good idea to modify `vars()` directly, since that's basically the same thing as `__dict__` hacking.